### PR TITLE
iptables: always use restore format

### PR DESCRIPTION
**Pull request by howardjohn**
This has been the default and only used value for >5 years. Just make it
the only option.

This introduces some breakages to tests, which are now legitimately
failing. I don't know why they didn't fail before. These are around the
new iptables idempotency logic. There are a few bugs here:

* NFLOG quoting was resulting in double quotes. This is due to a
  mismatch of iptables input vs iptables-save output (as indicated would
be a risk on the PR).
* We have some NFLOGs that are not on Istio chains which breaks the
  logic. We should have these! I removed them, but would loveto add them
back
* `-A PREROUTING -m conntrack --ctstate INVALID -j DROP` rule is not on
  Istio chain. We need to fix this.

### CODE REVIEW

**Review by leosarra — COMMENTED** [review]
I'm in favor of removing the non-restore format.
A lot of time has passed since the introduction of the restore format by default.
On top of that, I suspect no one is using the old format as I don't think there was an easy way to enable it via the charts.

File: tools/istio-iptables/pkg/log/command.go [hunk 1]
```diff
@@ -21,8 +21,6 @@ type Command struct {
 }
 
 var (
-	JumpInbound             = Command{"JumpInbound", "direct all traffic through ISTIO_INBOUND chain"}
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply tools/istio-iptables/pkg/log/command.go:24]: I don't actually want to, but the new idempotent logic doesn't allow istio rules in non-istip chains which this is.

[the remark this answers is not in the document]
Comment (leosarra) [reply tools/istio-iptables/pkg/log/command.go:24]: How does it not allow? Maybe that is not intentional and I can adjust it.

I was only aware of the warning I added in a non-istio chain that is not a jump rule to ISTIO_* 		https://github.com/istio/istio/blob/833fe1e8d884a3b59117d13734f0a409df11c295/tools/istio-iptables/pkg/builder/iptables_builder_impl.go#L79

but I don´t think the warning will be triggered for trace_loggings rules as those are added later as "extras" if TraceLogging is enabled
https://github.com/istio/istio/blob/833fe1e8d884a3b59117d13734f0a409df11c295/tools/istio-iptables/pkg/builder/iptables_builder_impl.go#L78-L92

**Comment by leosarra** [comment]
That may not be intentional, I will take a look tomorrow and I will update here with what can be done to bring those rules back

**Comment by leosarra** [comment]
Regarding drop-invalid...

By enabling restore format, the drop-invalid rule, which is not in an ISTIO chain (that is a problem by itself), is being executed as the first "commit" of the iptables-restore command.
```
* mangle
-A PREROUTING -m conntrack --ctstate INVALID -j DROP
COMMIT
* nat
-N ISTIO_INBOUND
-N ISTIO_REDIRECT
-N ISTIO_IN_REDIRECT
-N ISTIO_OUTPUT
...
COMMIT
```
This is a problem because the partial-atomicity of iptables-restore will still allow the `-A PREROUTING -m conntrack ...` rule to go through on reruns even though the commit on the *nat table will fail due to duplicate ISTIO chains.
Therefore, in the test, when the [third run](https://github.com/istio/istio/blob/master/tools/istio-iptables/pkg/capture/run_test.go#L412-L415) with `force-apply=true` is performed (for which it expects a failure), we end up with the expected failure but also two drop-invalid rules. The final-cleanup step of the test will only remove one, which later leads to an assert failure.

For this issue to occur, our iptables-restore input must meet two conditions:
- Changes to one table consist only of appending/inserting rules to default chains.
- These changes are placed in the input before any other table changes that create new chains.

I warned of this scenario in https://github.com/istio/istio/pull/50328#discussion_r1731548677 and that we should have kept it in mind for future iptables rules we add, but I failed to see that the config with drop-invalid was already satisfying all these conditions.
The test did not fail when the restore-format was disabled because the creation of iptables chains were the first commands to be executed in the old format.

For fixing this, we could do one of these three things:
1. Move the drop-invalid rule inside an Istio chain
2. Change the order in which we commit changes to tables: make sure that the first commit block of our iptables-restore always has a creation of an ISTIO chain.
3. Revert this part of the PR, but we would end up with a less backward compatible behavior of idempotency

I feel option 1. is the best approach, and perhaps 2. could be an extra ??.

Note: this affects also pre-idempotency-PR releases. istio-init reruns will append rules at every failed execution.

I think I found the issue for NFLOG as well, I will write a comment about that as well.

**Comment by leosarra** [comment]
>Now, regarding the NFLOGs issue.
>
>NFLOG quoting was resulting in double quotes. This is due to a
>mismatch of iptables input vs iptables-save output (as indicated would
>be a risk on the PR).


Yes, the rules that are applied are not necessarily the same as those you retrieve with iptables-save due to the inevitable round-trip to the kernel. However, I don't think this is where the issue is coming from.
The idempotency/reconciliation PR is resilient to this kernel's manipulation of the iptables input, as the iptables-save output is not parsed to the extent that it would lead to an issue.
The PR handling limits itself to counting the number of rules and checking the chains the rules belong to; all other arguments are ignored.

Also, I don't think the issue is related to the fact that we have NFLOG rules in non-ISTIO chains.

The iptables -C command that is failing after the restore format is generated from our original rule object, and it will undergo the same transformation by the kernel that the iptables -A command was subject to. Hence, having double-quotes, single-quotes, or no quotes in the iptables command does not matter, as they will all be treated the same.

This can be seen in this snippet:
```
$cat <<-EOF | iptables-restore
*nat
-N ISTIO_IN_REDIRECT
-A ISTIO_IN_REDIRECT -p tcp -j NFLOG --nflog-prefix "InboundCapture" --nflog-group 1337 --nflog-size 20
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
COMMIT
EOF

$iptables-save // Note that quotes are removed

# Generated by iptables-save v1.8.10 (nf_tables) on Wed Sep 11 13:30:46 2024
*nat
:ISTIO_IN_REDIRECT - [0:0]
-A ISTIO_IN_REDIRECT -p tcp -j NFLOG --nflog-prefix InboundCapture --nflog-group 1337 --nflog-size 20
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
COMMIT

// Verify that -C cmds exit with 0 regardless of the manipulation that the kernel performed to  --nflog-prefix

$sudo iptables -t nat -C ISTIO_IN_REDIRECT -p tcp -j NFLOG --nflog-prefix "InboundCapture" --nflog-group 1337 --nflog-size 20 // Exit 0
$sudo iptables -t nat -C ISTIO_IN_REDIRECT -p tcp -j NFLOG --nflog-prefix InboundCapture --nflog-group 1337 --nflog-size 20 // Exit 0
$sudo iptables -t nat -C ISTIO_IN_REDIRECT -p tcp -j NFLOG --nflog-prefix 'InboundCapture' --nflog-group 1337 --nflog-size 20 // Exit 0
```

The actual issue is caused by improper escaping in our rule-builder, which leads to two different rules depending on whether we use iptables-restore or go.exec:
- In our iptables-restore input:
iptables-restore will create 
-t nat -A ISTIO_IN_REDIRECT -p tcp -j NFLOG --nflog-prefix "InboundCapture" --nflog-group 1337 --nflog-size 20
- If we apply/check/delete via go.exec the rule we run is instead
-C|A|D ISTIO_IN_REDIRECT -p tcp -j NFLOG --nflog-prefix "\\"InboundCapture\\"" --nflog-group 1337 --nflog-size 20

These are two different rules with different meaning, so it makes sense that the iptables -C command returns an error in the idempotency check.
The escaping is missing in our iptables-restore input because of the strings.Join operation on the rule arguments that is performed when converting from Rule to string.
The issue is not reproducible if the restore-format is disabled because then the iptables append/delete/check operations are all aligned with the same (incorrect) input.
I think you've already encountered this issue, as you partially changed the rule builder here.

https://github.com/istio/istio/blob/e5e861b7d42ad7cd603108fef924ea0ae8eae6d2/tools/istio-iptables/pkg/builder/iptables_builder_impl.go#L127

but you need to change here as well
https://github.com/istio/istio/blob/e5e861b7d42ad7cd603108fef924ea0ae8eae6d2/tools/istio-iptables/pkg/builder/iptables_builder_impl.go#L87

After this the tests should pass even when those NFLOGs rules are appended to the non-istio chains.

**Comment by howardjohn** [comment]
Thanks for the comments! Super super help. I addresses the nflog one, pending on the INVALID one

> Move the drop-invalid rule inside an Istio chain

Seems reasonable, a bit annoying we need to create a new chain just to put 1 rule. Technically, we don't actually need to put the rule inside the new chain though :thinking: we just need to make a chain so it will fail later if it still exists...?

> Change the order in which we commit changes to tables: make sure that the first commit block of our iptables-restore always has a creation of an ISTIO chain.

This sounded good in theory but was thinking how to do it, and it would mean we order whether we create NAT or MANGLE or ... first which seems slightly odd...?

**Comment by leosarra** [comment]
> Thanks for the comments! Super super help.

I'm glad the comments came in handy :)

> Seems reasonable, a bit annoying we need to create a new chain just to put 1 rule. Technically, we don't actually need to put the rule inside the new chain though 🤔 we just need to make a chain so it will fail later if it still exists...?

Yes, in theory, we could create a chain in mangle that is never used and still keep the rule outside.
This will make iptables-restore fail if the chain already exists, and the partial atomicity will be enough to prevent duplicates of the drop-invalid rule.
On the other hand, having a chain without a rule will look weird. At that point, it may be worth just moving the rule there and making a jump, but this will introduce a small performance penalty due to the traffic going through an extra chain that did not exist before.
Honestly, I would prefer to move the rule inside the chain as that will make future CNI iptables reconciliation easier to perform.

> This sounded good in theory but was thinking how to do it, and it would mean we order whether we create NAT or MANGLE or ... first which seems slightly odd...?

Currently, the ordering in which the iptables-restore input is filled is a simple sorting on the keys of a map table_name:rules.
https://github.com/istio/istio/blob/e5e861b7d42ad7cd603108fef924ea0ae8eae6d2/tools/istio-iptables/pkg/builder/iptables_builder_impl.go#L366
We could make the sorting a bit more complex and prioritize tables in which we are creating a new chain
e.g:
```
sortedTables := maps.Keys(tableRulesMap)
slices.SortFunc(sortedTables, func(i, j string) int {
hasNewChainRule := func(rules []string) bool {
	for _, rule := range rules {
		if strings.Contains(rule, "-N") {
			return true
		}
	}
	return false
}
iHasNewChain := hasNewChainRule(tableRulesMap[i])
jHasNewChain := hasNewChainRule(tableRulesMap[j])
switch {
// Prioritize tables where we create new chains
case iHasNewChain && !jHasNewChain:
	return -1
case !iHasNewChain && jHasNewChain:
	return 1
// If both have new chain rules or both don't, sort by table name
default:
	return cmp.Compare(i, j)
}
})

for _, table := range sortedTables {
   ...
}
```

**Comment by bleggett** [comment]
> > Move the drop-invalid rule inside an Istio chain
> 
> Seems reasonable, a bit annoying we need to create a new chain just to put 1 rule. Technically, we don't actually need to put the rule inside the new chain though 🤔 we just need to make a chain so it will fail later if it still exists...?


I'm fine putting a single rule in an entirely bespoke chain - it's more readable, easier to reason about, and has nil perf impact.

As I think has been said before, any Istio rule outside of a named Istio chain is a bug. We should have 0 of them. Named chains are always the answer.

We should put that in a README somewhere.

**Comment by howardjohn** [comment]
> We should put that in a README somewhere.

Well now it will fail in CI at least :slightly_smiling_face: assuming the feature is added to the test

**Comment by howardjohn** [comment]
Sorry for the delay here. Updated invalid drop to use its own Istio chain

File: tools/istio-iptables/pkg/capture/run.go [hunk 2]
```diff
@@ -508,8 +509,8 @@ func (cfg *IptablesConfigurator) Run() error {
 
 	if redirectDNS {
 		// Jump from OUTPUT chain to ISTIOOUTPUT chain for all UDP traffic
-		cfg.ruleBuilder.AppendRule(iptableslog.JumpOutbound, constants.OUTPUT, constants.NAT, "-p", constants.UDP, "-j", constants.ISTIOOUTPUT)
-		cfg.ruleBuilder.AppendRule(iptableslog.JumpOutbound, constants.OUTPUT, constants.RAW, "-p", constants.UDP, "-j", constants.ISTIOOUTPUT)
+		cfg.ruleBuilder.AppendRule(iptableslog.UndefinedCommand, constants.OUTPUT, constants.NAT, "-p", constants.UDP, "-j", constants.ISTIOOUTPUT)
```
Comment (leosarra) [inline tools/istio-iptables/pkg/capture/run.go:512]: I think we should replace iptableslog.UndefinedCommand with iptables.JumpOutbound now that the nflog issue is addressed

Comment (howardjohn) [reply tools/istio-iptables/pkg/capture/run.go:512]: thanks, missed this one

File: cni/pkg/iptables/testdata/default_ipv6.golden [hunk 3]
```diff
@@ -1,48 +1,60 @@
...
+-N ISTIO_OUTPUT
+-A PREROUTING -j ISTIO_PRERT
+-A OUTPUT -j ISTIO_OUTPUT
+-A ISTIO_PRERT -m mark --mark 0x539/0xfff -j CONNMARK --set-xmark 0x111/0xfff
+-A ISTIO_OUTPUT -m connmark --mark 0x111/0xfff -j CONNMARK --restore-mark --nfmask 0xffffffff --ctmask 0xffffffff
+COMMIT
+* nat
+-N ISTIO_PRERT
+-N ISTIO_OUTPUT
+-A OUTPUT -j ISTIO_OUTPUT
```
Comment (bleggett) [inline cni/pkg/iptables/testdata/default_ipv6.golden:42]: hm, one minor annoyance is the new `golden` format makes it harder to distinguish which are V6 table rules and which are V4 table rules, and in general makes them less readable overall, which is a pity as they were a quick way to validate changes to logic/tests. Oh well, still better to ditch the option.

**Review by bleggett — APPROVED** [review]
:+1:

**Review by leosarra — APPROVED** [review]
Looking good. Thank you 👍

**Pull request merged**